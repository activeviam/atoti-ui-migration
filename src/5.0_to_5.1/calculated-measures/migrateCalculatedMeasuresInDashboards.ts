import {
  ContentRecord,
  CubeName,
  DashboardState,
  DataModel,
  deserializeDashboardState,
  MdxSelect,
  serializeDashboardState,
  WidgetWithQueryState,
} from "@activeviam/activeui-sdk-5.0";
import { migrateCalculatedMeasuresInMdx } from "./migrateCalculatedMeasuresInMdx";
import { produce } from "immer";
import _uniq from "lodash/uniq";

/**
 * Same as `migrateCalculatedMeasuresInWidgets` but for widgets within saved dashboards.
 */
export const migrateCalculatedMeasuresInDashboards = (
  dashboards: ContentRecord,
  dataModel: DataModel,
  namesOfCalculatedMeasurestoMigrate: string[],
): {
  measureToCubeMapping: { [measureName: string]: CubeName[] };
  migratedDashboards: ContentRecord;
} => {
  const measureToCubeMapping: { [measureName: string]: CubeName[] } = {};

  const migratedDashboards = produce(dashboards, (draft) => {
    // The children property is always defined for the `ui/dashboards` folder.
    const dashboardsContent = draft.children!.content.children ?? {};
    for (const dashboardId in dashboardsContent) {
      const dashboard: ContentRecord = dashboardsContent[dashboardId];
      const serializedDashboardState: DashboardState<"serialized"> = JSON.parse(
        dashboard.entry.content,
      );
      const deserializedDashboardState = deserializeDashboardState(
        serializedDashboardState,
      );

      const dashboardPages = deserializedDashboardState.pages;
      for (const pageId in dashboardPages) {
        const page = dashboardPages[pageId];
        const updatedWidgets: {
          [widgetId: string]: WidgetWithQueryState;
        } = produce(page.content, (draft) => {
          for (const widgetId in page.content) {
            const mdx: MdxSelect | undefined = page.content[widgetId].query.mdx;
            if (!mdx) {
              return;
            }
            const {
              migratedMdx,
              namesOfCalculatedMeasuresToMigrateInWidget,
              cubeName,
            } = migrateCalculatedMeasuresInMdx(
              mdx,
              namesOfCalculatedMeasurestoMigrate,
              dataModel,
            );

            namesOfCalculatedMeasuresToMigrateInWidget.forEach(
              (calculatedMeasureName) => {
                measureToCubeMapping[calculatedMeasureName] =
                  measureToCubeMapping[calculatedMeasureName]
                    ? _uniq([
                        ...measureToCubeMapping[calculatedMeasureName],
                        cubeName,
                      ])
                    : [cubeName];
              },
            );

            draft[widgetId].query.mdx = migratedMdx;
          }
        });
        dashboardPages[pageId].content = updatedWidgets;
      }

      dashboard.entry.content = JSON.stringify(
        serializeDashboardState(deserializedDashboardState),
      );
      dashboardsContent[dashboardId] = dashboard;
    }
    draft.children!.content.children = dashboardsContent;
  });
  return { measureToCubeMapping, migratedDashboards };
};
