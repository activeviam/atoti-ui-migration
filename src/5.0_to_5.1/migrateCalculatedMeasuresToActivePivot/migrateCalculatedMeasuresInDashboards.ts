import {
  ContentEntry,
  ContentRecord,
  CubeName,
  DashboardState,
  DataModel,
  MdxString,
  WidgetWithQueryState,
} from "@activeviam/activeui-sdk";
import { removeCalculatedMemberDefinitionFromMDXAndGetCubeName } from "./removeCalculatedMemberDefinitionFromMDXAndGetCubeName";
import { produce } from "immer";
import _merge from "lodash/merge";

interface DashboardsFolder {
  entry: ContentEntry;
  children: {
    thumbnails: ContentRecord;
    content: ContentRecord;
    structure: ContentRecord;
  };
}

/**
 * Takes a `ui/dashboards` folder from the content server, removes any calculated member definitions from the MDX and returns a `migratedDashboards` folder.
 * Also returns the cubeName for each calculated measure found.
 */
export const migrateCalculatedMeasuresInDashboards = (
  dashboards: DashboardsFolder,
  dataModel: DataModel,
  namesOfCalculatedMeasurestoMigrate: string[],
): {
  cubeNames: { [measureName: string]: CubeName };
  migratedDashboards: DashboardsFolder;
} => {
  const cubeNames: { [measureName: string]: CubeName } = {};

  const migratedDashboards = produce(dashboards, (draft) => {
    const allDashboards = dashboards.children.content.children ?? {};
    const updatedAllDashboards = produce(allDashboards, (draft) => {
      for (const dashboardId in allDashboards) {
        const dashboard: ContentRecord = allDashboards[dashboardId];
        const updatedDashboard = produce(dashboard, (draft) => {
          const dashboardState: DashboardState = JSON.parse(
            dashboard.entry.content,
          );
          const updatedDashboardState = produce(dashboardState, (draft) => {
            const dashboardPages = dashboardState.pages;
            const updatedDashboardPages = produce(dashboardPages, (draft) => {
              for (const pageId in dashboardPages) {
                const page = dashboardPages[pageId];
                const updatedWidget: {
                  [widgetId: string]: WidgetWithQueryState;
                } = produce(page.content, (draft) => {
                  for (const widgetId in page.content) {
                    const mdx: MdxString | undefined =
                      page.content[widgetId].query.mdx;
                    if (!mdx) {
                      return;
                    }
                    const {
                      stringifiedUpdatedMdx,
                      calculatedMeasuresWithCubeNames,
                    } = removeCalculatedMemberDefinitionFromMDXAndGetCubeName(
                      mdx,
                      namesOfCalculatedMeasurestoMigrate,
                      dataModel,
                    );
                    draft[widgetId].query.mdx = stringifiedUpdatedMdx;
                    _merge(cubeNames, calculatedMeasuresWithCubeNames);
                  }
                });
                draft[pageId].content = updatedWidget;
              }
            });
            draft.pages = updatedDashboardPages;
          });
          draft.entry.content = JSON.stringify(updatedDashboardState);
        });
        draft[dashboardId] = updatedDashboard;
      }
    });
    draft.children!.content.children = updatedAllDashboards;
  });
  return { cubeNames, migratedDashboards };
};
