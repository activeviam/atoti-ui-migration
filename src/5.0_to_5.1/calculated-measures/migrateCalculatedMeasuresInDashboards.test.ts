import { uiDashboardsFolder } from "../__test_resources__/uiDashboardsFolder";
import { migrateCalculatedMeasuresInDashboards } from "./migrateCalculatedMeasuresInDashboards";
import { sandboxDataModel } from "@activeviam/data-model-5.1/dist/__test_resources__";

const dataModels = { sandbox: sandboxDataModel };

// "pvSum ^ 2" is from cube "EquityDerivativesCubeDist", all others are from "EquityDerivativesCube".
const calculatedMeasureNames = [
  "Distinct count city",
  "Test calculated measure",
  "EXP pnl.Forex",
  "Log pv.SUM",
  "activeui5 calculated measure",
  "pvSum ^ 2",
];

describe("migrateCalculatedMeasuresInDashboards", () => {
  const { migratedDashboards, measureToCubeMapping } =
    migrateCalculatedMeasuresInDashboards(
      uiDashboardsFolder,
      dataModels,
      calculatedMeasureNames,
    );

  it("returns an object containing all the calculated measures used in the `ui/dashboards` folder with their corresponding cube name", () => {
    expect(measureToCubeMapping).toStrictEqual({
      "Distinct count city": ["EquityDerivativesCube"],
      "EXP pnl.Forex": ["EquityDerivativesCube"],
      "Log pv.SUM": ["EquityDerivativesCube"],
      "Test calculated measure": ["EquityDerivativesCube"],
      "activeui5 calculated measure": ["EquityDerivativesCube"],
      "pvSum ^ 2": ["EquityDerivativesCubeDist"],
    });
  });

  it("does not modify the content of an empty dashboard", () => {
    // "8c1" is an empty dashboard.
    expect(migratedDashboards.children!.content.children!["8c1"]).toStrictEqual(
      uiDashboardsFolder.children.content.children["8c1"],
    );

    // "a4f" is an empty dashboard inside a folder.
    expect(migratedDashboards.children!.content.children!["a4f"]).toStrictEqual(
      uiDashboardsFolder.children.content.children["a4f"],
    );
  });

  it("does not modify the content of a dashboard which does not contain calculated measures", () => {
    // "81a" is a dashboard with no calculated measures.
    expect(migratedDashboards.children!.content.children!["81a"]).toStrictEqual(
      uiDashboardsFolder.children.content.children["81a"],
    );

    // "e27" is a dashboard with no calculated measures inside a folder.
    expect(migratedDashboards.children!.content.children!["e27"]).toStrictEqual(
      uiDashboardsFolder.children.content.children["e27"],
    );
  });

  it("removes definitions of calculated measures saved in ActiveUI 5.0 from the `query.mdx` of all widgets from a dashboard with a single page", () => {
    // "b3e" is a dashboard with a single page, containing a single widget, containing a single calculated measure.
    expect(
      uiDashboardsFolder.children.content.children!["b3e"].entry.content,
    ).toContain("WITH  Member [Measures]");
    expect(
      migratedDashboards.children!.content.children!["b3e"].entry.content,
    ).not.toContain("WITH  Member [Measures]");

    // "c83" is a dashboard with a single page with multiple widgets, some of which contain calculated measures.
    expect(
      uiDashboardsFolder.children.content.children!["c83"].entry.content,
    ).toContain("WITH  Member [Measures]");
    expect(
      migratedDashboards.children!.content.children!["c83"].entry.content,
    ).not.toContain("WITH  Member [Measures]");
  });

  it("removes definitions of calculated measures saved in ActiveUI 5.0 from the `query.mdx` of all widgets from a dashboard with multiple pages", () => {
    // "ef0" is a dashboard with multiple pages, some of which contain calculated measures.
    expect(
      uiDashboardsFolder.children.content.children!["ef0"].entry.content,
    ).toContain("WITH  Member [Measures]");
    expect(
      migratedDashboards.children!.content.children!["ef0"].entry.content,
    ).not.toContain("WITH  Member [Measures]");
  });

  it("removes definitions of calculated measures saved in ActiveUI 5.0 from the `query.mdx` of all widgets from a dashboard with multiple pages inside a folder", () => {
    // "a18" is a dashboard inside a folder with multiple pages, some of which contain calculated measures.
    expect(
      uiDashboardsFolder.children.content.children!["a18"].entry.content,
    ).toContain("WITH  Member [Measures]");
    expect(
      migratedDashboards.children!.content.children!["a18"].entry.content,
    ).not.toContain("WITH  Member [Measures]");
  });

  it("does not remove the definition of a calculated measure which is not in the list `calculatedMeasureNames`", () => {
    // "a9e" is a dashboard containing 1 widget with a calculated measure which is not on the list of names of calculated measures to be migrated`.
    expect(migratedDashboards.children!.content.children!["a9e"]).toStrictEqual(
      uiDashboardsFolder.children.content.children["a9e"],
    );
  });

  it("removes definitions of calculated measures saved in ActiveUI 5.0 if their names are included in `calculatedMeasureNames` and does not remove them if they are not", () => {
    // "e54" is a dashboard containing 3 widgets.
    const migratedDashboardState = JSON.parse(
      migratedDashboards.children!.content.children!["e54"].entry.content,
    );
    const initialDashboardState = JSON.parse(
      uiDashboardsFolder.children.content.children!["e54"].entry.content,
    );

    // Widget "0" contains a calculated measure which is not on the list of `calculatedMeasureNames` ("Log City"), the MDX is unchanged.
    expect(migratedDashboardState.pages["p-0"].content["0"]).toStrictEqual(
      initialDashboardState.pages["p-0"].content["0"],
    );

    // Widget "1" contains a calculated measure on the list of `calculatedMeasureNames` ("Log pv.SUM") and another not on the list ("Log City").
    // The calculated measure definition is removed for ("Log pv.SUM") but not ("Log City").
    expect(
      migratedDashboardState.pages["p-0"].content["1"].query.mdx,
    ).toMatchInlineSnapshot(
      `"WITH  Member [Measures].[Log City] AS Log([Geography].[City], 10), FORMAT_STRING = \\"#,###.##\\"  SELECT NON EMPTY Hierarchize(Descendants({[Geography].[City].[AllMember]}, 1, SELF_AND_BEFORE)) ON ROWS, NON EMPTY {[Measures].[Log City], [Measures].[delta.SUM], [Measures].[Log pv.SUM]} ON COLUMNS FROM [EquityDerivativesCube] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS"`,
    );

    // Widget "2" contains only a calculated measure on the list of `calculatedMeasureNames`, the calculated measure definition is removed.
    expect(
      migratedDashboardState.pages["p-0"].content["2"].query.mdx,
    ).toMatchInlineSnapshot(
      `"SELECT NON EMPTY Hierarchize(Descendants({[CounterParty].[CounterParty].[AllMember]}, 1, SELF_AND_BEFORE)) ON ROWS, NON EMPTY {[Measures].[Log pv.SUM]} ON COLUMNS FROM [EquityDerivativesCube] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS"`,
    );
  });
});
