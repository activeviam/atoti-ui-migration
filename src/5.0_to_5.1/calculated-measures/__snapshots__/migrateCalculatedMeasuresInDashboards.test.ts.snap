// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`migrateCalculatedMeasuresInDashboards removes definitions of calculated measures saved in ActiveUI 5.0 from the \`query.mdx\` of all widgets from a dashboard with a single page 1`] = `
Object {
  "filters": Array [],
  "pages": Object {
    "p-0": Object {
      "content": Object {
        "0": Object {
          "mapping": Object {
            "columns": Array [],
            "measures": Array [
              "[Measures].[Distinct count city]",
            ],
            "rows": Array [
              "[Geography].[City].[City]",
            ],
          },
          "query": Object {
            "mdx": "SELECT NON EMPTY Hierarchize(Descendants({[Geography].[City].[AllMember]}, 1, SELF_AND_BEFORE)) ON ROWS, NON EMPTY {[Measures].[Distinct count city]} ON COLUMNS FROM [EquityDerivativesCube] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS",
            "updateMode": "once",
          },
          "serverKey": "Ranch 6.0",
          "widgetKey": "pivot-table",
        },
      },
      "layout": Object {
        "children": Array [
          Object {
            "leafKey": "0",
            "size": 1,
          },
        ],
        "direction": "row",
      },
      "name": "Page 1",
    },
  },
  "pagesOrder": Array [
    "p-0",
  ],
}
`;

exports[`migrateCalculatedMeasuresInDashboards removes definitions of calculated measures saved in ActiveUI 5.0 from the \`query.mdx\` of all widgets from a dashboard with a single page 2`] = `
Object {
  "filters": Array [],
  "pages": Object {
    "p-0": Object {
      "content": Object {
        "0": Object {
          "mapping": Object {
            "columns": Array [
              "ALL_MEASURES",
            ],
            "measures": Array [
              "[Measures].[Log pv.SUM]",
            ],
            "rows": Array [
              "[Underlyings].[Products].[ProductType]",
            ],
          },
          "query": Object {
            "mdx": "SELECT NON EMPTY {[Measures].[Log pv.SUM]} ON COLUMNS, NON EMPTY Hierarchize(Descendants({[Underlyings].[Products].[AllMember]}, 1, SELF_AND_BEFORE)) ON ROWS FROM [EquityDerivativesCube] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS",
            "updateMode": "once",
          },
          "serverKey": "Ranch 6.0",
          "widgetKey": "pivot-table",
        },
        "1": Object {
          "mapping": Object {
            "horizontalSubplots": Array [],
            "secondaryValues": Array [],
            "splitBy": Array [
              "ALL_MEASURES",
            ],
            "values": Array [
              "[Measures].[activeui5 calculated measure]",
            ],
            "verticalSubplots": Array [],
            "xAxis": Array [
              "[Currency].[Currency].[Currency]",
            ],
          },
          "query": Object {
            "mdx": "SELECT NON EMPTY Hierarchize(Descendants({[Currency].[Currency].[AllMember]}, 1, SELF_AND_BEFORE)) ON ROWS, NON EMPTY {[Measures].[activeui5 calculated measure]} ON COLUMNS FROM [EquityDerivativesCube] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS",
            "updateMode": "once",
          },
          "serverKey": "Ranch 6.0",
          "widgetKey": "plotly-line-chart",
        },
        "2": Object {
          "mapping": Object {
            "horizontalSubplots": Array [],
            "stackBy": Array [
              "ALL_MEASURES",
            ],
            "values": Array [
              "[Measures].[contributors.COUNT]",
            ],
            "verticalSubplots": Array [],
            "xAxis": Array [
              "[Time].[HistoricalDates].[AsOfDate]",
            ],
          },
          "query": Object {
            "mdx": "SELECT NON EMPTY [Time].[HistoricalDates].[AsOfDate].Members ON ROWS, NON EMPTY {[Measures].[contributors.COUNT]} ON COLUMNS FROM [EquityDerivativesCube] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS",
            "updateMode": "once",
          },
          "serverKey": "Ranch 6.0",
          "widgetKey": "plotly-stacked-column-chart",
        },
        "3": Object {
          "mapping": Object {
            "axis": Array [
              "[Geography].[City].[City]",
            ],
            "horizontalSubplots": Array [],
            "splitBy": Array [
              "ALL_MEASURES",
            ],
            "values": Array [
              "[Measures].[Distinct count city]",
              "[Measures].[contributors.COUNT]",
            ],
            "verticalSubplots": Array [],
          },
          "query": Object {
            "mdx": "SELECT NON EMPTY Hierarchize(Descendants({[Geography].[City].[AllMember]}, 1, SELF_AND_BEFORE)) ON ROWS, NON EMPTY {[Measures].[Distinct count city], [Measures].[contributors.COUNT]} ON COLUMNS FROM [EquityDerivativesCube] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS",
            "updateMode": "once",
          },
          "serverKey": "Ranch 6.0",
          "switchedTo": "plotly-clustered-bar-chart",
          "widgetKey": "plotly-radar-chart",
        },
      },
      "layout": Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "leafKey": "0",
                "size": 0.5,
              },
              Object {
                "leafKey": "2",
                "size": 0.5,
              },
            ],
            "direction": "column",
            "size": 0.5,
          },
          Object {
            "children": Array [
              Object {
                "leafKey": "1",
                "size": 0.5,
              },
              Object {
                "leafKey": "3",
                "size": 0.5,
              },
            ],
            "direction": "column",
            "size": 0.5,
          },
        ],
        "direction": "row",
      },
      "name": "Page 1",
    },
  },
  "pagesOrder": Array [
    "p-0",
  ],
}
`;

exports[`migrateCalculatedMeasuresInDashboards removes definitions of calculated measures saved in ActiveUI 5.0 from the \`query.mdx\` of all widgets from a dashboard with multiple pages 1`] = `
Object {
  "filters": Array [],
  "pages": Object {
    "p-0": Object {
      "content": Object {
        "0": Object {
          "mapping": Object {
            "horizontalSubplots": Array [],
            "secondaryValues": Array [],
            "splitBy": Array [
              "ALL_MEASURES",
            ],
            "values": Array [
              "[Measures].[Distinct count city]",
            ],
            "verticalSubplots": Array [],
            "xAxis": Array [
              "[Geography].[City].[City]",
            ],
          },
          "query": Object {
            "mdx": "SELECT NON EMPTY {[Measures].[Distinct count city]} ON COLUMNS, NON EMPTY Hierarchize(Descendants({[Geography].[City].[AllMember]}, 1, SELF_AND_BEFORE)) ON ROWS FROM [EquityDerivativesCube] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS",
            "updateMode": "once",
          },
          "serverKey": "Ranch 6.0",
          "widgetKey": "plotly-line-chart",
        },
      },
      "layout": Object {
        "children": Array [
          Object {
            "leafKey": "0",
            "size": 1,
          },
        ],
        "direction": "row",
      },
      "name": "Page 1",
    },
    "p-1": Object {
      "content": Object {
        "0": Object {
          "mapping": Object {
            "columns": Array [
              "ALL_MEASURES",
            ],
            "measures": Array [
              "[Measures].[Log pv.SUM]",
            ],
            "rows": Array [
              "[CounterParty].[CounterParty].[CounterPartyGroup]",
            ],
          },
          "query": Object {
            "mdx": "SELECT NON EMPTY Hierarchize(Descendants({[CounterParty].[CounterParty].[AllMember]}, 1, SELF_AND_BEFORE)) ON ROWS, NON EMPTY {[Measures].[Log pv.SUM]} ON COLUMNS FROM [EquityDerivativesCube] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS",
            "updateMode": "once",
          },
          "serverKey": "Ranch 6.0",
          "widgetKey": "pivot-table",
        },
        "1": Object {
          "mapping": Object {
            "horizontalSubplots": Array [],
            "stackBy": Array [
              "ALL_MEASURES",
            ],
            "values": Array [
              "[Measures].[gamma.SUM]",
            ],
            "verticalSubplots": Array [],
            "xAxis": Array [
              "[Booking].[Desk].[LegalEntity]",
            ],
          },
          "query": Object {
            "mdx": "SELECT NON EMPTY Hierarchize(Descendants({[Booking].[Desk].[AllMember]}, 1, SELF_AND_BEFORE)) ON ROWS, NON EMPTY {[Measures].[gamma.SUM]} ON COLUMNS FROM [EquityDerivativesCube] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS",
            "updateMode": "once",
          },
          "serverKey": "Ranch 6.0",
          "widgetKey": "plotly-stacked-column-chart",
        },
      },
      "layout": Object {
        "children": Array [
          Object {
            "leafKey": "0",
            "size": 0.5,
          },
          Object {
            "leafKey": "1",
            "size": 0.5,
          },
        ],
        "direction": "row",
      },
      "name": "Page 2",
    },
    "p-2": Object {
      "content": Object {
        "0": Object {
          "mapping": Object {
            "columns": Array [
              "ALL_MEASURES",
            ],
            "measures": Array [
              "[Measures].[pvSum ^ 2]",
            ],
            "rows": Array [
              "[Geography].[City].[City]",
            ],
          },
          "query": Object {
            "mdx": "SELECT NON EMPTY Hierarchize(Descendants({[Geography].[City].[AllMember]}, 1, SELF_AND_BEFORE)) ON ROWS, NON EMPTY {[Measures].[pvSum ^ 2]} ON COLUMNS FROM [EquityDerivativesCubeDist] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS",
            "updateMode": "once",
          },
          "serverKey": "Ranch 5.11",
          "widgetKey": "pivot-table",
        },
      },
      "layout": Object {
        "children": Array [
          Object {
            "leafKey": "0",
            "size": 1,
          },
        ],
        "direction": "row",
      },
      "name": "Page 3",
    },
    "p-3": Object {
      "content": Object {
        "0": Object {
          "mapping": Object {
            "columns": Array [],
            "measures": Array [
              "[Measures].[EXP pnl.Forex]",
            ],
            "rows": Array [
              "[Geography].[City].[City]",
            ],
          },
          "query": Object {
            "mdx": "SELECT NON EMPTY Hierarchize(Descendants({[Geography].[City].[AllMember]}, 1, SELF_AND_BEFORE)) ON ROWS, NON EMPTY {[Measures].[EXP pnl.Forex]} ON COLUMNS FROM [EquityDerivativesCube] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS",
            "updateMode": "once",
          },
          "serverKey": "Ranch 6.0",
          "widgetKey": "pivot-table",
        },
      },
      "layout": Object {
        "children": Array [
          Object {
            "leafKey": "0",
            "size": 1,
          },
        ],
        "direction": "row",
      },
      "name": "Page 4",
    },
  },
  "pagesOrder": Array [
    "p-0",
    "p-1",
    "p-2",
    "p-3",
  ],
}
`;

exports[`migrateCalculatedMeasuresInDashboards removes definitions of calculated measures saved in ActiveUI 5.0 from the \`query.mdx\` of all widgets from a dashboard with multiple pages inside a folder 1`] = `
Object {
  "filters": Array [],
  "pages": Object {
    "p-0": Object {
      "content": Object {
        "0": Object {
          "mapping": Object {
            "columns": Array [
              "ALL_MEASURES",
            ],
            "measures": Array [
              "[Measures].[Test calculated measure]",
            ],
            "rows": Array [
              "[Geography].[City].[City]",
              "[Currency].[Currency].[Currency]",
            ],
          },
          "query": Object {
            "mdx": "SELECT NON EMPTY Crossjoin(Hierarchize(Descendants({[Geography].[City].[AllMember]}, 1, SELF_AND_BEFORE)), Hierarchize(Descendants({[Currency].[Currency].[AllMember]}, 1, SELF_AND_BEFORE))) ON ROWS, NON EMPTY {[Measures].[Test calculated measure]} ON COLUMNS FROM [EquityDerivativesCube] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS",
            "updateMode": "once",
          },
          "serverKey": "Ranch 6.0",
          "widgetKey": "pivot-table",
        },
      },
      "layout": Object {
        "children": Array [
          Object {
            "leafKey": "0",
            "size": 1,
          },
        ],
        "direction": "row",
      },
      "name": "Page 1",
    },
    "p-1": Object {
      "content": Object {
        "0": Object {
          "mapping": Object {
            "columns": Array [
              "ALL_MEASURES",
            ],
            "measures": Array [
              "[Measures].[pvSum ^ 2]",
            ],
            "rows": Array [
              "[Underlyings].[Products].[ProductType]",
            ],
          },
          "query": Object {
            "mdx": "SELECT NON EMPTY Hierarchize(Descendants({[Underlyings].[Products].[AllMember]}, 1, SELF_AND_BEFORE)) ON ROWS, NON EMPTY {[Measures].[pvSum ^ 2]} ON COLUMNS FROM [EquityDerivativesCubeDist] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS",
            "updateMode": "once",
          },
          "serverKey": "Ranch 5.11",
          "widgetKey": "pivot-table",
        },
        "1": Object {
          "mapping": Object {
            "columns": Array [
              "ALL_MEASURES",
            ],
            "measures": Array [
              "[Measures].[pvSum ^ 2]",
            ],
            "rows": Array [
              "[Underlyings].[Products].[ProductType]",
            ],
          },
          "query": Object {
            "mdx": "SELECT NON EMPTY Hierarchize(Descendants({[Underlyings].[Products].[AllMember]}, 1, SELF_AND_BEFORE)) ON ROWS, NON EMPTY {[Measures].[pvSum ^ 2]} ON COLUMNS FROM [EquityDerivativesCubeDist] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS",
            "updateMode": "once",
          },
          "serverKey": "Ranch 5.11",
          "switchedTo": "kpi",
          "widgetKey": "pivot-table",
        },
      },
      "layout": Object {
        "children": Array [
          Object {
            "leafKey": "0",
            "size": 0.5,
          },
          Object {
            "leafKey": "1",
            "size": 0.5,
          },
        ],
        "direction": "row",
      },
      "name": "Page 2",
    },
    "p-2": Object {
      "content": Object {
        "0": Object {
          "mapping": Object {
            "columns": Array [
              "ALL_MEASURES",
            ],
            "measures": Array [
              "[Measures].[Log pv.SUM]",
            ],
            "rows": Array [
              "[Booking].[Desk].[LegalEntity]",
            ],
          },
          "query": Object {
            "mdx": "SELECT NON EMPTY Hierarchize(Descendants({[Booking].[Desk].[AllMember]}, 1, SELF_AND_BEFORE)) ON ROWS, NON EMPTY {[Measures].[Log pv.SUM]} ON COLUMNS FROM [EquityDerivativesCube] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS",
            "updateMode": "once",
          },
          "serverKey": "Ranch 6.0",
          "widgetKey": "pivot-table",
        },
        "1": Object {
          "mapping": Object {
            "columns": Array [
              "ALL_MEASURES",
            ],
            "measures": Array [
              "[Measures].[pvSum ^ 2]",
            ],
            "rows": Array [
              "[Underlyings].[Underlyings].[UnderlierType]",
            ],
          },
          "query": Object {
            "mdx": "SELECT NON EMPTY {[Measures].[pvSum ^ 2]} ON COLUMNS, NON EMPTY Hierarchize(Descendants({[Underlyings].[Underlyings].[AllMember]}, 1, SELF_AND_BEFORE)) ON ROWS FROM [EquityDerivativesCubeDist] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS",
            "updateMode": "once",
          },
          "serverKey": "Ranch 5.11",
          "widgetKey": "pivot-table",
        },
      },
      "layout": Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "leafKey": "0",
                "size": 0.5,
              },
              Object {
                "leafKey": "1",
                "size": 0.5,
              },
            ],
            "direction": "column",
            "size": 1,
          },
        ],
        "direction": "row",
      },
      "name": "Page 3",
    },
  },
  "pagesOrder": Array [
    "p-0",
    "p-1",
    "p-2",
  ],
}
`;

exports[`migrateCalculatedMeasuresInDashboards removes definitions of calculated measures saved in ActiveUI 5.0 from the \`query.mdx\` of widgets if their names are included in \`calculatedMeasureNames\` and does not remove them if they are not 1`] = `
Object {
  "filters": Array [],
  "pages": Object {
    "p-0": Object {
      "content": Object {
        "0": Object {
          "mapping": Object {
            "columns": Array [
              "ALL_MEASURES",
            ],
            "measures": Array [
              "[Measures].[Log City]",
              "[Measures].[delta.SUM]",
            ],
            "rows": Array [
              "[Geography].[City].[City]",
            ],
          },
          "query": Object {
            "mdx": "WITH  Member [Measures].[Log City] AS Log([Geography].[City], 10), FORMAT_STRING = \\"#,###.##\\"  SELECT NON EMPTY Hierarchize(Descendants({[Geography].[City].[AllMember]}, 1, SELF_AND_BEFORE)) ON ROWS, NON EMPTY {[Measures].[Log City], [Measures].[delta.SUM]} ON COLUMNS FROM [EquityDerivativesCube] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS",
            "updateMode": "once",
          },
          "serverKey": "Ranch 6.0",
          "widgetKey": "pivot-table",
        },
        "1": Object {
          "mapping": Object {
            "horizontalSubplots": Array [],
            "stackBy": Array [
              "ALL_MEASURES",
            ],
            "values": Array [
              "[Measures].[Log City]",
              "[Measures].[delta.SUM]",
              "[Measures].[Log pv.SUM]",
            ],
            "verticalSubplots": Array [],
            "xAxis": Array [
              "[Geography].[City].[City]",
            ],
          },
          "query": Object {
            "mdx": "WITH  Member [Measures].[Log City] AS Log([Geography].[City], 10), FORMAT_STRING = \\"#,###.##\\"  SELECT NON EMPTY Hierarchize(Descendants({[Geography].[City].[AllMember]}, 1, SELF_AND_BEFORE)) ON ROWS, NON EMPTY {[Measures].[Log City], [Measures].[delta.SUM], [Measures].[Log pv.SUM]} ON COLUMNS FROM [EquityDerivativesCube] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS",
            "updateMode": "once",
          },
          "serverKey": "Ranch 6.0",
          "widgetKey": "plotly-stacked-column-chart",
        },
        "2": Object {
          "mapping": Object {
            "horizontalSubplots": Array [],
            "sliceBy": Array [
              "[CounterParty].[CounterParty].[CounterPartyGroup]",
            ],
            "values": Array [
              "[Measures].[Log pv.SUM]",
            ],
            "verticalSubplots": Array [],
          },
          "query": Object {
            "mdx": "SELECT NON EMPTY Hierarchize(Descendants({[CounterParty].[CounterParty].[AllMember]}, 1, SELF_AND_BEFORE)) ON ROWS, NON EMPTY {[Measures].[Log pv.SUM]} ON COLUMNS FROM [EquityDerivativesCube] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS",
            "updateMode": "once",
          },
          "serverKey": "Ranch 6.0",
          "widgetKey": "plotly-pie-chart",
        },
      },
      "layout": Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "leafKey": "0",
                "size": 0.5,
              },
              Object {
                "leafKey": "2",
                "size": 0.5,
              },
            ],
            "direction": "column",
            "size": 0.5,
          },
          Object {
            "leafKey": "1",
            "size": 0.5,
          },
        ],
        "direction": "row",
      },
      "name": "Page 1",
    },
  },
  "pagesOrder": Array [
    "p-0",
  ],
}
`;
